catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/handlers/CqlTransformHandler.java

```diff
@@ -0,0 +1,224 @@
+      cqlRequest = mapper.readValue(body, CqlRequest.class);
+      if (cqlRequest == null) {
+        LOGGER.debug("Cql request parsed from body evaluated to  null.");
+        throw new NullPointerException("Cql request is null.");
+      }
+    } catch (Exception e) {
+      LOGGER.debug("Error fetching cql request");
+      response.status(HttpStatus.BAD_REQUEST_400);
+      return ImmutableMap.of("message", "Bad request");
+    }
+
+    Map<String, Serializable> arguments;
+
+    try {
+      arguments = mapper.readValue(body, Arguments.class).getSerializableArguments();
+    } catch (NullPointerException e) {
+      arguments = Collections.emptyMap();
+    }
+
+    LOGGER.trace("Finding transformer to transform query response.");
+
+    ServiceReference<QueryResponseTransformer> queryResponseTransformer =
+        queryResponseTransformers
+            .stream()
+            .filter(transformer -> transformerId.equals(transformer.getProperty("id")))
+            .findFirst()
+            .orElse(null);
+
+    if (queryResponseTransformer == null) {
+      LOGGER.debug("Could not find transformer with id: {}", transformerId);
+      response.status(HttpStatus.NOT_FOUND_404);
+      return ImmutableMap.of("message", "Service not found");
+    }
+
+    CqlQueryResponse cqlQueryResponse = util.executeCqlQuery(cqlRequest);
+
+    attachFileToResponse(request, response, queryResponseTransformer, cqlQueryResponse, arguments);
+
+    return "";
+  }
+
+  private void setHttpHeaders(Request request, Response response, BinaryContent content)
+      throws MimeTypeException, NullPointerException {
+    String mimeType = content.getMimeTypeValue();
+
+    if (mimeType == null) {
+      LOGGER.debug("Failure to fetch file extension, mime-type is empty");
+      throw new MimeTypeException("Binary Content contains null mime-type value.");
+    }
+
+    String fileExt = getFileExtFromMimeType(mimeType);
+
+    if (containsGzip(request)) {
+      LOGGER.trace("Request header accepts gzip");
+      response.header(HttpHeaders.CONTENT_ENCODING, GZIP);
+    }
+
+    response.type(mimeType);
+    String attachment =
+        String.format("attachment;filename=export-%s%s", Instant.now().toString(), fileExt);
+    response.header(HttpHeaders.CONTENT_DISPOSITION, attachment);
+  }
+
+  private String getFileExtFromMimeType(String mimeType)
+      throws MimeTypeException, NullPointerException {
+    MimeTypes allTypes = MimeTypes.getDefaultMimeTypes();
+    String fileExt = allTypes.forName(mimeType).getExtension();
+    if (fileExt == null || StringUtils.isEmpty(fileExt)) {
+      LOGGER.debug("Fetching file extension from mime-type resulted in null or empty.");
+      throw new NullPointerException("Failure fetching file extension from mime type");
+    }
+    return allTypes.forName(mimeType).getExtension();
+  }
+
+  private Boolean containsGzip(Request request) {
+    return request.headers(HttpHeaders.ACCEPT_ENCODING).toLowerCase().contains(GZIP);
```

Comment:
❓ Is it possible that there is no `ACCEPT_ENCODING` header in the request which would lead to an NPE here?

Reply:
there's a possibility. adding a NPE check

Reply:
You could do something like

```
Optional.ofNullable(request.headers(HttpHeaders.ACCEPT_ENCODING)).map(String::toLowerCase).orElse("").contains(GZIP);
```
which should avoid an explicit NPE check